[tune] Make the logging of the function API consistent and predictable

gehring opened the pull request:
## What do these changes do?

This is a re-implementation of the `FunctionRunner` which enforces some synchronicity between the thread running the training function and the thread running the Trainable which logs results. The main purpose is to make logging consistent across APIs in anticipation of a new function API which will be generator based (through `yield` statements). Without these changes, it will be impossible for the (possibly soon to be) deprecated reporter based API to behave the same as the generator based API.

This new implementation provides additional guarantees to prevent results from being dropped. This makes the logging behavior more intuitive and consistent with how results are handled in custom subclasses of Trainable.

New guarantees for the tune function API:

- Every reported result, i.e., `reporter(**kwargs)` calls, is forwarded to the appropriate loggers instead of being dropped if not enough time has elapsed since the last results.
- The wrapped function only runs if the `FunctionRunner` expects a result, i.e., when `FunctionRunner._train()` has been called. This removes the possibility that a result will be generated by the function but never logged.
- The wrapped function is not called until the first `_train()` call. Currently, the wrapped function is started during the setup phase which could result in dropped results if the trial is cancelled between `_setup()` and the first `_train()` call.
- Exceptions raised by the wrapped function won't be propagated until all results are logged to prevent dropped results.
- The thread running the wrapped function is explicitly stopped when the `FunctionRunner` is stopped with `_stop()`.
- If the wrapped function terminates without reporting `done=True`, a duplicate result with `{"done": True}`, is reported to explicitly terminate the trial, but will not be logged.

## Related issue number

#3956
#3949
#3834

gehring commented:
@richardliaw Would you mind checking this out. I didn't want to go ahead with the generator based API until I could guarantee that both the old and new function runner would behave mostly the same.

This implementation seems to work as expected. Most test pass at the moment with the exception of the tests which expect the last result to be repeated when `done=True` is never set. I'll go through and fix the tests once you've given me your thoughts on the new logging behavior and runner implementation.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11754/
Test FAILed.

richardliaw commented:
Sorry for the wait, I'll get to this today! (And feel free to ping if you don't hear)

richardliaw commented:
jenkins retest this please

richardliaw commented:
I think it looks good! What do you think about having an option that
toggles asynchronous vs synchronous behavior for the reporter?

On Mon, Feb 11, 2019 at 12:27 PM Robert Nishihara <notifications@github.com>
wrote:

> Assigned #4011 <https://github.com/ray-project/ray/pull/4011> to
> @richardliaw <https://github.com/richardliaw>.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/4011#event-2131697973>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEUc5VoBpd0W_rGjxn1OcGBBN8gtCb9nks5vMdIxgaJpZM4azZg_>
> .
>

andrewztan commented on python/ray/tune/trainable.py:
```diff
@@ -440,12 +440,17 @@ def _export_model(self, export_formats, export_dir):
+        def _trainable_func(self, config, reporter):
+            output = train_func(config, reporter)
+            reporter(done=True)
+            return output
```
In order for this to be compatible with existing trainable functions, I believe `_trainable_func` would need to return a function rather than the output of the called function. Are you proposing an alternative to wrapping functions as Trainable objects for logging?

andrewztan commented on python/ray/tune/trainable.py:
```diff
@@ -440,12 +440,17 @@ def _export_model(self, export_formats, export_dir):
+        def _trainable_func(self, config, reporter):
+            output = train_func(config, reporter)
+            reporter(done=True)
+            return output
```
In order for this to be compatible with existing trainable functions, I believe `_trainable_func` would need to return a function rather than the output of the called function. Are you proposing an alternative to wrapping functions as Trainable objects for logging?

gehring replied:
The `_trainable_func` method belongs to the `FunctionWrapper` class. As I understand it, it is not part of the `Trainable` class API. In both the old and new implementation, this method is used to dynamically create a `Trainable` class which uses the provided function.

This change is only to simplify how functions are wrapped within a Trainable object. I felt it was clunky and unnecessary to have to call this method "twice", e.g., `_trainable_func()(config, reporter)`. Alternatively, if we would want to keep it such that it returns a function, we could consider making it a property (but then again, this feels unnecessary).

Additionally, I am getting ready to propose a generator based function API which would simplify things greatly. With this new way of wrapping the function, backwards compatibility should be clean and straightforward.

This new implementation of ray's FunctionRunner accounts for these change so shouldn't cause any problems. This should not affect any older versions unless someone tries to unpickle incompatible versions or tries to access the "private" method directly. Even then, I would think it unlikely as the FunctionWrapper instances aren't exposed through the public API and users should only be using one version of ray at a time.

That being said, this change is minor and completely optional. We can revert back to the old way but I am not convinced this is necessary. Given the re-write of `FunctionWrapper`, this would be a good time to change it.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/11878/
Test FAILed.

gehring commented:
> What do you think about having an option that toggles asynchronous vs synchronous behavior for the reporter?

This would be fairly straightforward to do but I would highly recommend against as I don't see any advantages in doing so.  Adding the option would add complexity to the API (and a source of confusion) while not offering any tangible benefits. In principal, most of the computation time should be spent within the training function so, in the best case, I don't expect any significant performance gains to offset the added complexity (and source of bugs).

While the old version was asynchronous it was not parallel (due to the GIL). Making them fully asynchronous again would only provide performance benefits if logging had especially long blocking calls.

Given this change is only to provide consistency between the Trainable API, the new generator function API and the reporter function API, I'm not keen on allowing users to opt out of the predictable behavior but I'm open to arguments for asynchronicity. Even more so, if the reporter function API is to be deprecated in favor of the generator function API and is kept around only for backwards compatibility purposes.

(As a related aside, any idea how long python 2.7 will be supported by ray? An `async`/`await` implementation of the function runner would have been very compact and understandable but I couldn't use it because of the need for backward compatibility.)

gehring commented:
> jenkins retest this please

@richardliaw There are several tests that would require changing before they can be passed as they assume some weird corner case behavior. The new implementation avoids these cases by always reporting `done=True`. Pretty much all failing tests seemed to use this corner cases as a way to avoid manually reporting `done=True` and should be easy to fix. Once I've addressed all your concerns with the `FunctionWrapper`, I'll go through and add the changes to the tests to this pull request.

Alternatively, if returning a simple result, `{"done": True}`, is not desirable. We could remove the requirement that the `FunctionWrapper` returns done. I've only kept it in because there was a special block of code in the old version dedicated to preventing this from happening, going as far a repeating the last result to avoid it.

richardliaw commented:
> In principal, most of the computation time should be spent within the training function so, in the best case, I don't expect any significant performance gains to offset the added complexity (and source of bugs).

That's a good point. 

> any idea how long python 2.7 will be supported by ray?

No clue... although I think the broader community would be really interested in this discussion (enough to warrant a post on the dev list I think!)

> Alternatively, if returning a simple result, `{"done": True}`, is not desirable.

I think returning a simple result, `{"done": True}` is fine. It would probably be good to add a comment or log something to tell the user that a `done=True` result is being provided, and may affect result output (specifically for users that have their own logging mechanisms).

We should also add test we don't report `done=True` twice, in the case the user sets `done=True` themselves.

richardliaw commented on python/ray/tune/function_runner.py:
```diff
@@ -40,101 +49,212 @@ def __call__(self, **kwargs):
 
     def _stop(self):
-        self._status_reporter._stop()
+        # notify the status reporter to stop if still running
```
in the spirit of simplicity, do we actually need any of this?

gehring replied:
Good catch! I added that before I had completely committed to synchronizing the `FunctionRunner` and the thread. I ended up keeping it as an undocumented way for the function to catch the stop (with the `should_stop` method I had).

Given this isn't really part of the API and not really supported, the whole thing should definitely get stripped.

richardliaw commented on python/ray/tune/function_runner.py:
```diff
@@ -40,101 +49,212 @@ def __call__(self, **kwargs):
+            logger.debug((
+                    "Thread runner raised StopIteration. Interperting it as a "
+                    "signal to terminate the thread without error."))
+            pass
```
this `pass` isn't needed right?

richardliaw commented on python/ray/tune/function_runner.py:
```diff
@@ -40,101 +49,212 @@ def __call__(self, **kwargs):
+        self._continue_semaphore.acquire()
+
+        # check if the stop signal was sent during the semaphore wait
+        if self.should_stop():
```
do we need `should_stop`?

gehring replied:
Nope! See other related comment.

gehring commented:
> At a high level, this change of behavior is good; thanks for making this PR!
> 
> Overall, I would prefer removing some complexity (i.e., removing code paths like `should_stop` and cleanups from `_stop`, as these have not been issues in the past - this is a pretty small edge case and people will probably checkpoint their stuff with some frequency anyways).
> 
> Let me know.

I agree with you. It should all be stripped it away now. Let me know if there is anything else. I should be able to get around to fixing the tests in the next couple of days.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12025/
Test FAILed.

richardliaw commented:
Thanks! I think this looks good and would be ready after the tests are added.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12029/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12574/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12585/
Test FAILed.

richardliaw commented on python/ray/tune/tests/test_trial_runner.py:
```diff
@@ -413,30 +413,10 @@ def train(config, reporter):
         self.assertEqual(trial.last_result[TIMESTEPS_TOTAL], 99)
 
-    def testNoRaiseFlag(self):
-        def train(config, reporter):
```
is there some way to keep this?

gehring replied:
I'm not sure I see what it should be testing since the function runner will always return at least one result, i.e., `{"done": True}`. Let me know what you have in mind and I will be happy to put it back.

richardliaw commented on python/ray/tune/tests/test_trial_runner.py:
```diff
@@ -376,7 +376,7 @@ def train(config, reporter):
         })
         self.assertEqual(trial.status, Trial.TERMINATED)
-        self.assertEqual(trial.last_result[TIMESTEPS_TOTAL], 100)
+        self.assertEqual(trial.last_result[DONE], True)
```
`self.assertEqual(trial.last_result[TIMESTEPS_TOTAL], 100)` should also be true right?

gehring replied:
I believe so. Though that behavior is already check for in the cleaned up `testReportTimeStep`.

There is a bit of overlap between `testReportTimeStep`, `testAbruptReturn`, and `testNoDoneReceived`. We could probably combine `testAbruptReturn` and `testNoDoneReceived`, and leave `testReportTimeStep` to check for the timestep counting behavior. What do you think?

richardliaw replied:
Yeah I think that's fair - can you combine those two?

gehring replied:
Done

richardliaw commented on python/ray/tune/function_runner.py:
```diff
@@ -41,105 +47,184 @@ def __call__(self, **kwargs):
 
     def _train(self):
-        time.sleep(
-            self.config.get("script_min_iter_time_s",
```
nit: `script_min_iter_time_s` is no longer a relevant parameter anywhere in the codebase right? Would it be possible to remove? Otherwise we can just make an issue and address a cleanup later.

gehring replied:
Do you mean remove it from the tests since it is not longer in the function runner? If so, I just clean the test a bit more and removed it.

gehring replied:
I've removed all of them found in test_trial_runner.py. If there are any others hiding around, I'm for the idea of creating a separate issue.

richardliaw commented on python/ray/tests/test_multi_node.py:
```diff
@@ -426,7 +426,7 @@ def test_run_driver_twice(ray_start_regular):
     for i in range(2):
         time.sleep(0.1)
-        reporter(timesteps_total=i, mean_accuracy=i+97)  # report metrics
+        reporter(timesteps_this_iter=i, mean_accuracy=i+97)  # report metrics
```
this doesn't need to be changed right?

gehring replied:
Sorry, this change is wrong. Given it increments the timesteps_total manually, this should be reverted. See my other reply about timesteps_total and timesteps_this_iter.

richardliaw commented on python/ray/tune/examples/genetic_example.py:
```diff
@@ -24,7 +24,7 @@ def michalewicz_function(config, reporter):
 
     # Negate y since we want to minimize y value
-    reporter(timesteps_total=1, neg_mean_loss=-y)
+    reporter(timesteps_this_iter=1, neg_mean_loss=-y)
```
this doesn't need to be changed right?

gehring replied:
Sorry, the old function runner seemed to have an odd behavior where timesteps_total was converted to timesteps_this_iter before being returned. I just went around and changed all timesteps_total using the function API as a precaution. Given this runs only for one iteration, this can be reverted.

richardliaw commented on python/ray/tune/function_runner.py:
```diff
@@ -41,105 +47,184 @@ def __call__(self, **kwargs):
-                res.setdefault(TRAINING_ITERATION, self._iteration)
-            return res
+        # time per iteration is recorded directly in the reporter to ensure
+        # any delays in logging results aren't counted
```
can we do `kwargs.copy()` just in case?

richardliaw replied:
(sorry, should have caught this in an earlier review)

gehring replied:
I'm fairly certain python create a new dict at every call, but, given the cost of copying should be low and somewhat infrequent, I've added it for piece of mind.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12637/
Test FAILed.

gehring commented:
> Thanks for cleaning this up! Just one thing -- @gehring, in `test_trial_runner.TrialRunnerTest`, can you add a test using the function API that makes sure all values reported are observed?

@richardliaw I've added test to check for all values reported. I also added a method to check that the same results are returned through a Trainable class as through the function API. Also added a test to ensure that only a single done is returned.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12723/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12724/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12725/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12729/
Test FAILed.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12768/
Test FAILed.

gehring commented:
> Awesome! Left small comments; it would be good to get the tests to pass now and then we can merge.

@richardliaw The tests seem to pass on my end now. Let me know if you have any other remaining comments.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12803/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12809/
Test FAILed.

richardliaw commented:
I was just debugging a Jenkins issue and I realized that there are two problems if the user doesn't report `done = i == 99`, which is rather cumbersome and a common failure mode:

1. one user problem that this introduces is that `trial.last_result` becomes effectively useless 

2. `on_trial_complete` no longer receives anything useful; which would break a lot of current search algorithms.

One workaround is for the function_runner to call `reporter(__done__=True)`. This will not be exposed to the user, and will not be passed to any TrialSchedulers or SearchAlgorithms. Instead, there could be some handling to make sure the previous result is passed into `on_trial_complete`. Let me know what you think, and I can try to implement it if necessary.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12822/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12837/
Test FAILed.

gehring commented:
> I was just debugging a Jenkins issue and I realized that there are two problems if the user doesn't report `done = i == 99`, which is rather cumbersome and a common failure mode:
> 
> 1. one user problem that this introduces is that `trial.last_result` becomes effectively useless
> 2. `on_trial_complete` no longer receives anything useful; which would break a lot of current search algorithms.
> 
> One workaround is for the function_runner to call `reporter(**last_result, __done__=True)`. This result will not be exposed to the user, and will skip the logging, but will be passed into `on_trial_complete` for the scheduler and search algorithms. Let me know what you think, and I can try to implement it if necessary.
> 
> EDIT: I've made a PR against your fork, let me know what you think

@richardliaw Thanks for the PR. I won't be able to look at it closely for a day or two, but here are my high level thoughts.

I think that a `reporter(..., __done__=True)` could work fine but I think given there is the possibility of changing the API to a generator based API (i.e., without a reporter), we might just be kicking the problem further down the road. Additionally, it seems to me that it would require reimplementing some of the behavior provided by `Trial` (or some other class) for keeping track of the last result, editing it when necessary. It might be best to try to find a more general solution.

Without having thought through the details, I'm thinking using the standard `StopIteration` exception might solve this problem with little effort. If we allow `Trainable._train()` and `Trainable.train()` to raise a `StopIteration` exception, we would only need to modify the class in charge of manipulating the last results (not sure on the top of my head if that is `Trainable`, `Trial`, or one of the runner/executors). 

The logic to do this would look like this:

```python
class ClassThatHandlesTrainCalls(object):

    ...

    def function_responsible_for_last_result(self, ...):
        # any necessary work to do before the next iteration call
        ...
        try:
            self._last_results = self._trianable.train().copy()
        except StopIteration:
            self._last_results[DONE] = True
        # whatever else this is responsible for updating
        ...
```

A `StopIteration` solution would make the implementation of generator based API very simple with the added benefit of allowing custom `Trainable` classes to implement their own termination mechanics by raising `StopIteration`. Plus, we avoid having to manage last results in several places in the code base, dodging future issues with bugs and inconsistencies. The only issue with this approach I could foresee would come the fact that the logic is distributed through ray. What do you think?

richardliaw commented:
Let me know when you take a look at the PR!

The main notions I'm sticking with are the following:

1. Whatever we do needs to be very simple for users; I would like as much
as possible to avoid any new concepts or gotchas (i.e., StopIteration,
remembering to set Done, etc -- this is not great IMO), especially in
places like Trainable where it doesn't make sense to raise StopIteration.
My PR against your branch explicitly avoids this without requiring to
reimplement any of the other parts of Tune.
2. We can do away with the "reporter" specifically, but keep in mind,
backwards compat is important at least for a while after introducing a new
"recommended" API. There is another API that we are considering internally
that does not *require* `reporter` nor a generator-based API. The high
level idea is to use some global context within each worker. We will push a
PR soon, and I'd love to get your feedback then.

Thanks,
Richard

On Wed, Mar 13, 2019 at 1:38 PM gehring <notifications@github.com> wrote:

> I was just debugging a Jenkins issue and I realized that there are two
> problems if the user doesn't report done = i == 99, which is rather
> cumbersome and a common failure mode:
>
>    1. one user problem that this introduces is that trial.last_result
>    becomes effectively useless
>    2. on_trial_complete no longer receives anything useful; which would
>    break a lot of current search algorithms.
>
> One workaround is for the function_runner to call reporter(**last_result,
> __done__=True). This result will not be exposed to the user, and will
> skip the logging, but will be passed into on_trial_complete for the
> scheduler and search algorithms. Let me know what you think, and I can try
> to implement it if necessary.
>
> EDIT: I've made a PR against your fork, let me know what you think
>
> @richardliaw <https://github.com/richardliaw> Thanks for the PR. I won't
> be able to look at it closely for a day or two, but here are my high level
> thoughts.
>
> I think that a reporter(..., __done__=True) could work fine but I think
> given there is the possibility of changing the API to a generator based API
> (i.e., without a reporter), we might just be kicking the problem further
> down the road. Additionally, it seems to me that it would require
> reimplementing some of the behavior provided by Trial (or some other
> class) for keeping track of the last result, editing it when necessary. It
> might be best to try to find a more general solution.
>
> Without having thought through the details, I'm thinking using the
> standard StopIteration exception might solve this problem with little
> effort. If we allow Trainable._train() and Trainable.train() to raise a
> StopIteration exception, we would only need to modify the class in charge
> of manipulating the last results (not sure on the top of my head if that is
> Trainable, Trial, or one of the runner/executors).
>
> The logic to do this would look like this:
>
> class ClassThatHandlesTrainCalls(object):
>
>     ...
>
>     def function_responsible_for_last_result(self, ...):
>         # any necessary work to do before the next iteration call
>         ...
>         try:
>             self._last_results = self._trianable.train().copy()
>         except StopIteration:
>             self._last_results[DONE] = True
>         # whatever else this is responsible for updating
>         ...
>
> A StopIteration solution would make the implementation of generator based
> API very simple with the added benefit of allowing custom Trainable
> classes to implement their own termination mechanics by raising
> StopIteration. Plus, we avoid having to manage last results in several
> places in the code base, dodging future issues with bugs and
> inconsistencies. The only issue with this approach I could foresee would
> come the fact that the logic is distributed through ray. What do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/4011#issuecomment-472595892>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEUc5dyc7cG-tw-7YV5yE-M0Vfj6Qlo9ks5vWWHVgaJpZM4azZg_>
> .
>

richardliaw commented:
This now looks good to me and I think we can merge, although it'd be great to get your eyes on some of the minor implementation changes that I've made.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12921/
Test FAILed.

gehring commented:
@richardliaw 

Now that this is blocking another PR, I am happy merging your `__duplicate__` approach and finally merging this PR. Though, I do have some follow up for the `StopIteration` discussion. We could open a new issue/PR to continue this discussion and avoid delaying this PR further.

> 1. Whatever we do needs to be very simple for users; I would like as much
as possible to avoid any new concepts or gotchas (i.e., StopIteration,
remembering to set Done, etc -- this is not great IMO), especially in
places like Trainable where it doesn't make sense to raise StopIteration.
My PR against your branch explicitly avoids this without requiring to
reimplement any of the other parts of Tune.

I wasn't suggesting that the user is meant to raise `StopIteration`, simply that its use is supported. More generally, I think the logic supporting trials to terminate on their own should be independent from the logic in charge of the results.

To be clear, I think the current `{"done": True, ...}` API is fine and can definitely stay the encouraged /public way of terminating trials from within. However, fully supporting unexpected trial termination would give a lot more flexibility both to the public and private API. In a way, this is already supported since regular exceptions need to be caught and reported so all is needed is a way to trigger this without propagating an error. Support for `StopIteration` would simply involve treating those particular exceptions the same way python control flow does.

Supporting `StopIteration` would allow for a very straightforward function runner:

```python
class FunctionRunnerV2(Trainable):
    def _setup(self, config):
        if is_reporter_required(self._train_func):
            # create a backwards compatible generator
       else:
           self._train_iter = self._train_func(config)

    def _train(self):
        # raises StopIteration when exhausted, i.e., when the training function terminates/returns
        return next(self._train_iter)
```

Finally, supporting `StopIteration` should be quite simple. It would probably only require the following changes to `TrialRunner`:
```python
class TrialRunner(object):
    # some other methods
    ...
    def _process_trial(self, trial):
        try:
            result = self.trial_executor.fetch_result(trial)
            # process result
            ...
        # extra code for StopIteration
        except StopIteration:
            # trial is exhausted, do any necessary bookkeeping 
            trial.set_last_result_done(True)
            ...
        except Exception:
            # process other exceptions normally
            logger.exception("Error processing event.")
            ...
```

> 2. We can do away with the "reporter" specifically, but keep in mind,
backwards compat is important at least for a while after introducing a new
"recommended" API. There is another API that we are considering internally
that does not *require* `reporter` nor a generator-based API. The high
level idea is to use some global context within each worker. We will push a
PR soon, and I'd love to get your feedback then.

Any reason to prefer a global context over an iterator/generator? I'm usually a little apprehensive when relying on some global state, but I'm excited to see what you guys are working on! Do you have a design I can check out?

As for backwards compat, I am definitely keeping it in mind! I don't think it would be too complicated to maintain it.

gehring commented:
> This now looks good to me and I think we can merge, although it'd be great to get your eyes on some of the minor implementation changes that I've made.

Just went through, as long as the `__duplicate__` approach hasn't broken any test then everything looks all good to me!

richardliaw commented:
> I wasn't suggesting that the user is meant to raise StopIteration, simply that its use is supported. More generally, I think the logic supporting trials to terminate on their own should be independent from the logic in charge of the results.

OK I agree with this and also agree it shouldn't be hard to add. Instead of a FunctionRunnerV2, I'd maybe create a IteratorRunner and have a check if the input is an IteratorRunner. Sorry for the miscommunication! Let's open up a new issue

> Any reason to prefer a global context over an iterator/generator? I'm usually a little apprehensive when relying on some global state, but I'm excited to see what you guys are working on! Do you have a design I can check out?

Yes! It's a light impl actually but maybe @noahgolmant can comment (or open a new Issue as RFC).

> Just went through, as long as the __duplicate__ approach hasn't broken any test then everything looks all good to me!

Thanks a bunch! I'll fix up the tests and get this merged; I think this will clean up a lot. Thanks!

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12987/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12994/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/12995/
Test PASSed.

richardliaw commented:
Thanks a bunch for this contribution @gehring!

gehring commented:
@richardliaw No problem! Ray/tune/rllib is a great set of libraries and I am happy to contribute!
